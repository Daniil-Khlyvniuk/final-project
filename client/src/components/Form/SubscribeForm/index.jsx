import React, { useState } from 'react'
import { Formik, Form, Field } from 'formik'
import { Button, Alert } from '@mui/material'
import { SUBSCRIBE_SCHEMA } from '../setting/Schemes'
import TextInput from '../setting/TextInput'
import useSnack from '../../../utils/customHooks/useSnack'
import { addSubscribe } from '../../../utils/API/subscribersAPI'
import { styled } from '@mui/material/styles'

const StyledForm = styled(Form)(() => ({
	display: 'flex',
	width: '100%',
}))

const StyledAlert = styled(Alert)(() => ({
	width: '100%',
}))

const SubscribeForm = () => {
	const {handleSnack} = useSnack()
	const [subscribeStatus, setSubscribeStatus] = useState(null)
	const handleSubmit = async ({ email }, formikFunctions) => {
		try {
			const res = await addSubscribe(email)
			if (res.status === 200) {
				setSubscribeStatus({ success: 'You successfully subscribed!' })
				handleSnack({message: 'You successfully subscribed', style: 'success'})
			}
		}
		catch (er) {
			setSubscribeStatus({ error: er.response.data.message })
			handleSnack({message: er.response.data.message, style: 'warning'})
		}
		formikFunctions.resetForm()
		setTimeout(() => setSubscribeStatus(null), 5000)
	}

	return (
		<Formik
			initialValues={{ email: '' }}
			validationSchema={SUBSCRIBE_SCHEMA}
			onSubmit={handleSubmit}
		>
			{(formikProps) => (
				<>
					<StyledForm noValidate>
						<Field
							component={TextInput}
							type="email"
							placeholder="e-mail"
							name="email"
							onBlur={formikProps.handleBlur}
							onChange={formikProps.handleChange}
							error={true}
							asyncborderradius={'on'}
							fullWidth={true}
						/>

						<Button
							type='submit'
							variant="contained"
							disabled={
								!formikProps.isValid ||
								formikProps.isSubmitting
							}
							asyncborderradius={'on'}
							sx={{
								paddingLeft: '40px',
								paddingRight: '40px',
							}}
						>
							send
						</Button>
					</StyledForm>

					{!formikProps.isValid && (
						<StyledAlert
							severity="error"
							icon={false}
						>
							{formikProps.errors.email}
						</StyledAlert>
					)}
					{subscribeStatus && subscribeStatus['success'] && (
						<StyledAlert
							icon={false}
							severity="success"
						>
							{subscribeStatus.success}
						</StyledAlert>
					)}
					{subscribeStatus && subscribeStatus['error'] && (
						<StyledAlert
							icon={false}
							severity="success"
						>
							{subscribeStatus.error}
						</StyledAlert>
					)}
				</>
			)}
		</Formik>
	)
}

export default SubscribeForm
