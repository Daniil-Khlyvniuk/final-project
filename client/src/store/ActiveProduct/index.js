
export { default as activeProductSelector } from './selectors'
export { default as activeProductOperations } from './operations'
import activeProductReducer from './activeProductSlice'

export default activeProductReducer