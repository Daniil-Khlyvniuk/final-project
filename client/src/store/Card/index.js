export { default as cardSelectors } from './selectors'
export { default as cardOperations } from './operations'
import * as cardActions from './cardSlice'

export default cardActions


