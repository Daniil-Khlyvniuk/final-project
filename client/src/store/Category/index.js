export { default as categorySelectors } from './selectors'
export { default as categoryOperations } from './operations'

import * as categoryActions from './categorySlice'
export default categoryActions