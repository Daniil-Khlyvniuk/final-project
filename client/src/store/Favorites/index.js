export { default as favoritesSelectors } from './selectors'
export { default as favoritesOperations } from './operations'

import * as favoritesActions from './favoritesSlice'
export default favoritesActions