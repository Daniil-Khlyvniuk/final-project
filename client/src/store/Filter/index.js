import reducer from './filterSlice'

export {default as filterSelectors} from './selectors'
export {default as filterOperations} from './operations'

export default reducer