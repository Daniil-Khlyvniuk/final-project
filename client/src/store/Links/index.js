export {default as linksSelectors} from './selectors'
export {default as linksOperations} from './operations'

import linksReducer from './linksSlice'
export default linksReducer