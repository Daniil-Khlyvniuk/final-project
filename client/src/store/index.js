import { configureStore } from '@reduxjs/toolkit'
import productsSlice from './Products/productsSlice'
import linksSlice from './Links/'
import categorySlice from './Category/categorySlice'
import sliderSlice from './Slider'
import modalSlice from './Modal/modalSlice'
import shoppingBagSlice from './ShoppingBag/shoppingBagSlice'
import productSlice from './Product/productSlice'
import filterSlice from './Filter'

import userSlice from './User'

const store = configureStore({
	reducer: {
		slides: sliderSlice,
		products: productsSlice,
		links: linksSlice,
		category: categorySlice,
		modal: modalSlice,
		shoppingBag: shoppingBagSlice,
		user: userSlice,
		product : productSlice,
		filter: filterSlice,

	},
	middleware: (getDefaultMiddleware) =>
		getDefaultMiddleware({
			serializableCheck: {
				// Ignore these action types
				ignoredActions: ['modal/modalToggle', 'user/setToken', 'user/fetchUser/pending','user/fetchUser/fulfilled','filter/setFiltersFromQueryString'],
			},
		}),
})

export default store
