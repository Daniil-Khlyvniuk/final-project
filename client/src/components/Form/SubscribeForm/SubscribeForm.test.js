import React from 'react'
import SubscribeForm from './index'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from '../../../store'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

describe('Smoke test SubscribeForm', () => {
	const setUp = () => {
		render(
			<Provider store={store}>
				<BrowserRouter>
					<SubscribeForm />
				</BrowserRouter>
			</Provider>
		)
	}

	let container
	beforeEach(() => {
		container = setUp()
	})
	it('should render SubscribeForm', () => {
		expect(container)
	})
	it('SubscribeForm Snapshot test', () => {
		expect(container).toMatchSnapshot()
	})
})

describe("Testing formik on SubscribeForm", () => {
	it('should rendering and submitting a basic Formik form', () => {
		const handleSubmit = jest.fn()

		render(
			<Provider store={store}>
				<BrowserRouter>
					<SubscribeForm />
				</BrowserRouter>
			</Provider>
		)

		userEvent.type(screen.getByTestId('email'), 'john.dee@someemail.com')

		userEvent.click(screen.getByTestId('button', { type: /submit/i }))

		waitFor(() =>
			expect(handleSubmit).toHaveBeenCalledWith({
				email: 'john.dee@someemail.com',
			}),
		)
	})
})