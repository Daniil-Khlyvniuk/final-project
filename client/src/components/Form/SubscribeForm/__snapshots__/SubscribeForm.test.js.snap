// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Smoke test SubscribeForm SubscribeForm Snapshot test 1`] = `undefined`;
